### ISSUE TITLE
treemacs and lsp conflict via imenu-create-index-list

### ISSUE THREAD

**Issue by emacs18** [Alexander-Miller/treemacs#451]
Treemacs and lsp are both great packages.  Unfortunately both cannot coexist because they both want to set imenu-create-index-list.  Treemacs does this to show tags from say GNU global tag table, and lsp does this to show information from lsp backend.  This issue was first reported in spacemacs as https://github.com/syl20bnr/spacemacs/issues/12384

I bring this to your attention in case someone might come up with a solution to let treemacs and lsp co-exist.

**Comment by duianto** [comment]
Let's ping @yyoncho he's on the lsp-mode team, and would surely like to get this fixed as well.

**Comment by yyoncho** [comment]
I am already looking into it but I am unable to figure out what the issue is. As far as I can see treemacs is only a consumer of the imenu items and does not alter it. @emacs18 can you point me to the offending code or provide steps to reproduce?

**Comment by emacs18** [comment]
I wrote the following in `org-mode' then translated into text. Sorry I
don't use github often enough that I haven't learned markdown.

The steps to see the problem are detailed below.

- Enable both `lsp' and `treemacs' layers in spacemacs.

- Add the following to `c++-mode-hook' per instruction on treemacs page:

  ,----
  | (setq-local imenu-create-index-function #'ggtags-build-imenu-index)
  `----

- Set `debug-on-variable-change' to `imenu-create-index-function' after
  starting emacs so that emacs will be stopped when
  `imenu-create-index-function' is modified.

- Load a C++ file. Stack trace is shown showing `setq-local' line above
  which is part of my own `c++-mode-hook'.

- Continue emacs stopped by hitting the breakpoint above. It will now
  hit the breakpoint again. This time `lsp' hook is sets
  `imenu-create-index-function' to some other value. Stack trace is
  showing where I chopped off long lines:

  ,----
  |   Debugger entered--setting imenu-create-index-function in buffer layerops.cpp to ggtags-build-imenu-index: 
  |     debug--implement-debug-watch(imenu-create-index-function ggtags-build-imenu-index set #<buffer layerops.cpp>)
  |     make-local-variable(imenu-create-index-function)
  |     lsp-enable-imenu()
  |     lsp--text-document-did-open()
  |     lsp--open-in-workspace(#s(lsp--workspace :parser #s(lsp--parser :headers nil :body ...
  |     #f(compiled-function (response) #<bytecode 0x29c4a91>)(#<hash-table equal 1/65 0x4281f45>)
  |     #f(compiled-function () #<bytecode 0x2983649>)()
  |     #f(compiled-function (result) #<bytecode 0x2983661>)(#<hash-table equal 1/65 0x4281f45>)
  |     lsp--parser-on-message(#s(lsp--parser :headers nil :body nil :reading-body nil ...
  |     #f(compiled-function (proc output) #<bytecode 0x29d20f5>)(#<process ccls> ...
  `----

Thus the value of `imenu-create-index-function' is
`lsp--imenu-create-index' rather than `ggtags-build-imenu-index'.
Obviously `treemacs' will have difficulty showing tags from GNU global
table.

**Comment by yyoncho** [comment]
> Obviously treemacs' will have difficulty showing tags from GNU global table.

As far as I can see this is not the case - treemacs will show the data that comes from lsp-mode which generally is considered more accurate since it is always up to date. Note that imenu format is common and treemacs will be able to show the imenu tags no matter whether imenu-create-index-function is coming from ggtags or lsp-mode.

**Comment by duianto** [comment]
@emacs18 Regarding Github markdown.

Inline code, is marked with a single backtick before and after.
`` ` ``(message "test")`` ` `` results in: `(message "test")`

A multi line code block is marked with three backticks on the line above the first line and on the line below the last line.
https://help.github.com/en/articles/basic-writing-and-formatting-syntax#quoting-code

It's also possible to add language specific syntax highlighting by adding the language name after the first three backticks.
https://help.github.com/en/articles/creating-and-highlighting-code-blocks

#### For example:
`` ```elisp ``
(defun paste-at-end-of-line ()
  "Paste at the end of the current line.
With a space before the pasted text."
  (interactive)
  (goto-char (line-end-position))
  (insert (concat " " (car kill-ring))))
`` ``` ``
#### results in:
```elisp
(defun paste-at-end-of-line ()
  "Paste at the end of the current line.
With a space before the pasted text."
  (interactive)
  (goto-char (line-end-position))
  (insert (concat " " (car kill-ring))))
```

**Comment by Alexander-Miller** [comment]
> As far as I can see this is not the case - treemacs will show the data that comes from lsp-mode which generally is considered more accurate since it is always up to date. Note that imenu format is common and treemacs will be able to show the imenu tags no matter whether imenu-create-index-function is coming from ggtags or lsp-mode.

This. 

I don't see what this issue has to do with treemacs, or even lsp. Treemacs doesn't care what else you have running, or where those tags come from, it just reads the tags of whatever provider is around, be it lsp or tags or a regex scan, the resulting format is always the same.

> Obviously treemacs' will have difficulty showing tags from GNU global table.

That's because *you* decided to you lsp-mode, so you'll see tags coming from your language server, not your tags table. So where's the problem?

**Comment by emacs18** [comment]
Thank you all for looking into this.  I agree that my earlier analysis was incorrect.
It appears that the root cause of my problem was corrupt compile_commands.json file.
This seemed to have confused lsp which in turn prevented treemacs from working correctly.

I misread treemacs documentation and thought that I needed to set imenu-create-index-function
in order to enable tags feature.  Now I know that tags feature can also use information from lsp
which is even better.

I don't use lsp much, so I did not realize that it was not working due to bad compile_commands.json.
After fixing this, lsp works well and also treemacs.  Sorry for the noise.

Also thanks to @duianto for pointing out markdown features.
I jotted down these in my org notes for future reference.
